src/View/BookingCodes.php

```diff
@@ -1,12 +1,343 @@
+    }
+
+ 	/**
+	 * Next Booking Codes by Email Cron event for timeframe.
+     * based on current timestamp calculated start and period 
+	 *
+	 * @param Timestamp $tsInitial
+	 * @param int $periodMonths
+     * 
+     * @return DateTime   Datetime of nex Cron Event.
+	 */
+    public static function nextCronEmailEvent($tsInitial,$periodMonths): DateTimeImmutable {
+        $start = new DateTimeImmutable("@{$tsInitial}");
+        if($tsInitial >= strtotime("today"))
+            return $start;
+
+        $end = new DateTime("today");
+        $diff = $start->diff($end);
+        
+        $yearsInMonths = $diff->format('%r%y') * 12;
+        $months = $diff->format('%r%m');
+        $totalMonths = $yearsInMonths + $months;
+        $numPeriods=floor($totalMonths / $periodMonths) + 1;        
+        $fullMonth=$numPeriods * $periodMonths;
+
+        $dtTo=$start->modify("midnight last day of next month +" . ($fullMonth - 1) . " month");
+        
+        $daydiff=$dtTo->format("j") - $start->format("j");
+        if($daydiff > 0 ) {
+          $dtNextCronEvent=$dtTo->modify("-" . $daydiff . " days");
+        } 
+        else 
+           $dtNextCronEvent=$dtTo;
+        
+        return $dtNextCronEvent;
+                
+    }
+
+ 	/**
+	 * CMB2 callback on field saved
+	 *
+     * @param bool              $updated Whether the metadata update action occurred.
+     * @param string            $action  Action performed. Could be "repeatable", "updated", or "removed".
+     * @param CMB2_Field object $field   This field object
+ 	 */
+    public static function cronEmailCodesSaved( $updated, $action, $field): void {
+        if(!$updated || empty($field->object_id())) return;
+            
+        switch($action) {
+            case 'updated':
+                if(empty(@$field->value()['cron-booking-codes-enabled']))
+                    delete_post_meta($field->object_id(), self::$nextCronEmailID);
+                else
+                {
+                    $dtNextCron=self::nextCronEmailEvent(@$field->value()['cron-email-booking-code-start'], @$field->value()['cron-email-booking-code-nummonth']);
+                    update_post_meta( $field->object_id(), self::$nextCronEmailID, $dtNextCron->getTimestamp() ); 
+                }
+                break;
+
+            case 'removed':
+                delete_post_meta($field->object_id(), self::$nextCronEmailID);
+                break;
+
+            default:
+                break;
+
+        }
+    }
+    
+ 	/**
+	 * CMB2 sanitize field callback
+	 *
+     * @param  mixed      $value      The unsanitized value from the form.
+     * @param  array      $field_args Array of field arguments.
+     * @param  CMB2_Field $field      The field object
+     *
+     * @return array                  Sanitized value to be stored.
+ 	 */
+    public static function sanitizeCronEmailCodes( $value, $field_args, $field ): array {
```

Comment:
Wäre es vielleicht ne Möglichkeit auch aus dem Timeframe Model ein DateTime Objekt zu ziehen? Oder in dem BookingCode Modell ein Datetime objekt zu beziehen?